### ISSUE TITLE
Share chainload toolchain with triplets / build libraries with the  CMAKE_TOOLCHAIN_FILE settings

### ISSUE THREAD

**Issue by jonathanverner** [microsoft/vcpkg#27445]
****

**Is your feature request related to a problem? Please describe.**
In our usecase, triplets are checked-in to the repository and, hence, common to all environments. However,
our toolchain files/locations (in particular, cross-compiling toolchain files) will typically depend on the environment.
Currently, it seems that the chainload toolchain needs to be specified by the triplet, so this means each developer
(and buildbot) needs to have the toolchain installed in the exact same location or patch the triplet before building.

**Proposed solution**
If we could pass the chainload toolchain location from the main CMakeLists.txt to the triplet 
(or. even better, use the chainload toolchain specified by the main CMakeLists.txt or on the commandline when building
libraries --- this makes much more sense, but is backward incompatible). 

**Describe alternatives you've considered**
Patching the triplets before building. This is frustrating, though.

**Context**
It says in the documentation that the toolchain can be shared (https://vcpkg.readthedocs.io/en/latest/users/buildsystems/cmake-integration/#using-multiple-toolchain-files), but doesn't say how.

**Comment by Neumann-A** [comment]
> , so this means each developer (and buildbot) needs to have the toolchain installed in the exact same location 

You can use `${CMAKE_CURRENT_LIST_DIR}` in the triplet file. So store the toolchain somewhere relative to the triplet and you are good to go (See e.g. https://github.com/Neumann-A/my-vcpkg-triplets/blob/master/x64-win-llvm.cmake)

**Comment by jonathanverner** [comment]
> > , so this means each developer (and buildbot) needs to have the toolchain installed in the exact same location
> 
> You can use `${CMAKE_CURRENT_LIST_DIR}` in the triplet file. So store the toolchain somewhere relative to the triplet and you are good to go (See e.g. https://github.com/Neumann-A/my-vcpkg-triplets/blob/master/x64-win-llvm.cmake)

Except that the toolchain will typically contain environment dependent things (compiler paths, sdk locations, etc.), so we don't really want to commit it to VCS. So there is really no "relative" path to put it...

More generally, when building, one wants to use the same toolchain for building the code as for building the dependencies. 
So it doesn't seem right to specify the toolchain in the triplet, which only applies to the dependencies.

**Comment by Neumann-A** [comment]
> So it doesn't seem right to specify the toolchain in the triplet, which only applies to the dependencies.

Then reuse the toolchain from the dependencies for your build? What is stopping you from doing that?

> Except that the toolchain will typically contain environment dependent things (compiler paths, sdk locations, etc.), so we don't really want to commit it to VCS

Don't use absolute paths. Use `find_(program|file)` to control the toolchain via typically CMake means. Your toolchain really shouldn't have absolute installation paths.

**Comment by jonathanverner** [comment]
> > So it doesn't seem right to specify the toolchain in the triplet, which only applies to the dependencies.
> 
> Then reuse the toolchain from the dependencies for your build? What is stopping you from doing that?
> 

Nothing really, just that it doesn't feel right :-) Also I want to keep as close to standard CMake as possible even when using vcpkg. I.e. I would like the CMAKE_TOOLCHAIN_FILE to behave the same way as it does when not using vcpkg. (I already need to hack around the fact that vpckg appropriates this variable for itself, but I can do that...)

> > Except that the toolchain will typically contain environment dependent things (compiler paths, sdk locations, etc.), so we don't really want to commit it to VCS
> 
> Don't use absolute paths. Use `find_(program|file)` to control the toolchain via typically CMake means. Your toolchain really shouldn't have absolute installation paths.

Using `find_program` is nice when you have toolchains which are located in predictable places (e.g. installed from system packages). However, when the toolchain is installed by the developers into arbitrary folders, this doesn't work very well. Of course, I can work around by mandating a specific location where the toolchains will be looked for, but it does remove flexibility.

**Comment by jonathanverner** [comment]
> > (I already need to hack around the fact that vpckg appropriates this variable for itself, but I can do that...)
> 
> `include(vcpkg.cmake)` in `CMAKE_TOOLCHAIN_FILE` is also possible instead of setting `CMAKE_TOOLCHAIN_FILE ` to `vcpkg.cmake`.

Yes, but then the toolchain (which specifies the compiler, not the dependency handler) is tied to vcpkg. However, as I said, this I can work-around. I.e. my `CMakeLists.txt` looks for `vcpkg` and sets `CMAKE_TOOLCHAIN_FILE` appropriately, before calling `project`.

> > However, when the toolchain is installed by the developers into arbitrary folders, this doesn't work very well.
> 
> Define an environment variable for the base installation path and look for that. Error if it is not set. (Or let people set `CMAKE_PROGRAM_PATH` as a requirement?)

Yes, this is probably the way to go. However, I wanted to get as close as possible to the following developer experience:

1. A developer, that doesn't have (doesn't want to have) any custom settings, just wants to compile the
project:

```
git clone repo
mkdir build
cd build
cmake ../repo
```

2. A developer, who has his own custom toolchain file which he wants to use:
```
git clone repo
mkdir build
cd build
cmake -DCMAKE_TOOLCHAIN_FILE=/home/user/mytoolchain.cmake ../repo
```

Currently, I don't see a way to get there. In particular, there is no way around hardcoding the set of available toolchains by specifying them in the triplets.
